File: pa1/List.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */


    //-----------------------------------------------------------------------------
//Programmer: Abbas Engineer
//CruzID  :   afengine
//Project :   PA1
//Version :   1.0
//Class   : CMPS 101 Prof. Tantalo
//-----------------------------------------------------------------------------
public class List {
       
    private class Node{
      // Fields
      int data;
      Node next;
      Node prev;
      
      // Constructor
      Node(int data) { this.data = data; next = null; prev = null; }
      
      // toString():  overrides Object's toString() method
      public String toString() { 
         return String.valueOf(data); 
      }
      
      // equals(): overrides Object's equals() method
      public boolean equals(Object x){
         boolean eq = false;
         Node that;
         if(x instanceof Node){
            that = (Node) x;
            eq = (this.data==that.data);
         }
         return eq;
      }
    
}
// Fields
   private Node front;
   private Node back;
   private int  length;
   private int  index;
   private Node cursor;

   // Constructor
   List() { 
      front = back = cursor = null; 
      length = 0; 
      index = -1;
   }
   
   
   //------------------------------ Access Functions -------------------------//

// isEmpty()     Returns true if this List is empty, false otherwise.
   boolean isEmpty() { 
      return length<=0; 
   }
   
// when the cursor is between the bounds of 0 and length -1 returns true
   boolean indexIs(){
       return index <= length - 1 && index >= 0;
   }

// getLength()   Returns length of this List.
   int length() { 
      return length; 
   }
 
// If cursor is defined, returns the index of the cursor element,
// otherwise returns -1.
   int index(){
   
       if (cursor == null )
            return -1;
        else 
            return index;
   }

// Front()       Returns front element.   // Pre: !this.isEmpty()
   int front(){
      if( this.isEmpty() ){
         throw new RuntimeException
        ("List Error: getFront() called on empty List");
      }
      return front.data;
   }

// int Back()    Returns back element.   // Pre: !this.isEmpty()
   int back(){
      if( this.isEmpty() ){
         throw new RuntimeException(
            "List Error: getBack() called on empty List");
      }
      return back.data;
   }
   
// int get()     Returns cursor element. // Pre: length()>0, index()>=0
   int get(){
   if(this.isEmpty() ){
       throw new RuntimeException(
       "List Error: get() called on empty List");
   }
   if(cursor == null){
       throw new RuntimeException(
       "Index Error: get() called an undefined cursor");
   }
   return cursor.data;
   }

// Returns true if and only if this List and L are the same
// integer sequence. The states of the cursors in the two Lists
// are not used in determining equality.
   public boolean equals(Object x){
      boolean eq  = false;
      List L;
      Node N, M;

      if(x instanceof List){
         L = (List)x;
         N = this.front;
         M = L.front;
         eq = (this.length==L.length);
         while( eq && N!=null ){
            eq = N.equals(M);
            N = N.next;
            M = M.next;
         }
      }
      return eq;
   }
   
//----------------------- Manipulation Functions-----------------------------//
// Resets this List to its original empty state.  
    void clear() {
length = 0;
index  = -1;
front = back = null ;
}

// If List is non-empty, places the cursor under the front element,
// otherwise does nothing.
    void moveFront() {
    if (! this.isEmpty() ){
    cursor = front ;
    index = 0;
    }
    else 
        ;
    }

// If List is non-empty, places the cursor under the back element,
// otherwise does nothing.
    void moveBack() {
    if (! this.isEmpty() ){
    cursor = back ;
    index = length - 1 ;
    }
    else
        ;
    }
        
// If cursor is defined and not at front, moves cursor one step toward
// front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is undefined does nothing.        
    void movePrev(){
        if (this.indexIs() ){
            cursor = cursor.prev;
            index--;
        }
        else{
            
            cursor = null ;
            index = -1;
        }
        
    }

// If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing.
    void moveNext(){
    if (this.indexIs() ){
            cursor = cursor.next;
            index++;
        }
    else {
        index = -1 ;
        }  

    } 

// Insert new element into this List. If List is non-empty,
    void prepend(int data) {
        Node N = new Node(data);
        if (this.isEmpty() ){
            front = back = N;
        }

        else if(front != null){

               N.next = front;
               N.prev = null;
               front.prev = N;
               front = N;
           
               index++;
        }
        length++;
        
    }

// insertion takes place before front element.
// Insert new element into this List. If List is non-empty,
    void append(int data) {
      Node N = new Node(data);
      if( this.isEmpty() ) { 
         front = back = N;
      }else{ 
         
         back.next = N; 
         N.prev = back;
         back = N; 
         N.next = null;
         
      }
      length++;
   }

    void insertBefore(int data){
        Node N = new Node(data);
        if(this.isEmpty() || cursor == front){
            prepend(data);
          
        }
        else if(!this.indexIs() ){
            throw new RuntimeException(
            "Index Error: insertBefore() called on undefined cursor");
        }
        else{
           Node prev = cursor.prev;
            //System.out.println("in else 1");
           
           N.next = cursor;
           N.prev = prev;
           prev.next = N;
           cursor.prev = N;
           //System.out.println("out of else ");
        length++;
        index++;
           //System.out.println("in else last");
            } 
         
      
    }

// Pre: length()>0, index()>=0
// Inserts new element after cursor.
    void insertAfter(int data){
        Node N = new Node(data);
        if (this.isEmpty() ){
            prepend(data);
        }
        else if( ! this.indexIs() ){
            throw new RuntimeException(
            "Index Error: insertAfter() called on undefined cursor");
        }
        else if(cursor == back){
            append(data);
        }
        else{
            Node next = cursor.next;
            
            N.next = cursor.next;
            N.prev = next;
            next.prev = N;
            cursor.next = N;
            length++;
        }
        
        
    } 
 
// Pre: length()>0, index()>=0
// Deletes the front element. Pre: length()>0            
    void deleteFront(){
        if(this.isEmpty() ){
         throw new RuntimeException(
            "List Error: deleteFront() called on empty List");
      }
       
      if(! this.isEmpty() ){
         front = front.next;
         if(front != null )
         front.prev = null;
         if (index != -1){
             index--; 
         }
	
         
      }else if (this.length() == 1){
         front = back = null;
         index = -1;
      }
      length--;
    }

// Deletes the back element. Pre: length()>0
    void deleteBack() {
        if(this.isEmpty() ){
         throw new RuntimeException(
            "List Error: deleteBack() called on empty List");
      }
      if(! this.isEmpty() ){
         if(cursor == back){
             index = -1;
         }
         back = back.prev;
         if(back != null){
         back.next = null;
         }

         
         length--;
         
      }else if (this.length() == 1){
         front = back = null;
         index =-1;
         
      }
      
    }
    
 // Deletes cursor element, making cursor undefined.
 // Pre: length()>0, index()>=0
    void delete() {
        if(this.isEmpty() || ! this.indexIs() ){
         throw new RuntimeException(
            "List Error: delete() called on empty List or cursor out of index");
      }
        else{
        if (cursor.next != null) { 
            cursor.next.prev = cursor.prev; 
        } 
        if (cursor.prev != null) { 
            cursor.prev.next = cursor.next; 
        } 
           cursor = null;
        }
        index = -1;
        length --;
    }

//------------------------------ Other Functions -----------------------------//
 
 // Overrides Object's toString method. Returns a String
 // representation of this List consisting of a space
 // separated sequence of integers, with front on left.
    public String toString() {
        StringBuffer sb = new StringBuffer();
      Node N = front;
      while(N!=null){
         // System.out.println("tostring1");
         sb.append(N.toString());
         sb.append(" ");
         N = N.next;
         // System.out.println("tostring2");
      }
      return new String(sb);
    }
  


 // copy()
 // Returns a new List identical to this List.
    List copy(){
      List L = new List();
      Node N = this.front;

      while( N!=null ){
         L.append(N.data);
         N = N.next;
      }
      return L;
    }
}

File: Graph-C/backup/List.c
/* Programmer: Abbas Engineer
     Cruz ID : Afengine
     Version : 1.0
  Description: Graph Shortest Path and Distance using BFS
                List.c
  Prof. Tantalo PA4


*/

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// struct is same as public class Node in java
typedef struct NodeObj
{
    int data;
    struct NodeObj* next;
    struct NodeObj* prev;

}NodeObj;

typedef NodeObj* Node; // private Node type

typedef struct ListObj
{
    Node front;
    Node back;
    int length;
    Node cursor;
    int index;
}ListObj;


Node newNode(int data)
{
  //  Node N = NULL;
    Node N = malloc(sizeof(NodeObj));
    N->data = data;
    N->next = NULL;
    N->prev = NULL;
    return (N);
}


void freeNode(Node* pN)
{
    if (pN != NULL && *pN != NULL)
    {
        free(*pN);
        *pN = NULL;
    }
}

// newList()
// equivalent to List class constructor
List newList(void)
{
    //List L = NULL;
   List L = malloc(sizeof(ListObj));
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->length = 0;
    L->index = -1;

    return (L);
}

bool notEmpty(List L){
if(length(L)>0) {return true;}
return false;
}

bool indexNotNull(List L){
if(index(L) != -1) {return true;}
return false;
}

bool cFront(List L){
if(L->cursor == L->front){return true;}
else{
    return false;
}
}

// freeList()
// frees all heap memory associated with List *pL
// and sets *pL to NULL
void freeList(List* pL)
{
    //printf("List.c: freeList() called\n");
    if (pL != NULL && *pL!= NULL)
    {

      while (notEmpty(*pL))
      {
          //printf("List.c: freeList(): deleteFront() called\n");
          clear(*pL);
      }
      //printf("!!List.c: freeList(): free(*pL) called for List\n");
      free(*pL);
      *pL = NULL;
    }
}

// Access functions ---------------------------------------

// length()
// returns the number of elements in this List
int length(List L) {

    if( L==NULL ){
      printf("List error: List already empty \n");
      exit(1);
   }
    return L->length;
}

// index()
// returns the index of the cursor of the list
int index(List L)
{
    //printf("List.c: index() is called.\n");
    if (L == NULL)
    {
        printf("List Error: calling index() on NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
        return -1;
    }
    return L->index;
}

// front()
// returns the front element. pre: length() > 0
int front(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling front() on NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
        printf("List: front() Empty List \n");
        return -1;
    }
    return L->front->data;
}

// back()
// returns the back element. pre: length() > 0
int back(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling back() on NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
        printf("List: back() Empty List \n");
        return -1;
    }
    return L->back->data;
}

// get()
// returns the cursor element. pre: length() > 0 and index() > 0
int get(List L) {
   if( !notEmpty(L) ){
      printf("List error: List already empty \n");
      exit(1);
   }
   if(L == NULL){
    printf("List error: Null List Reference");
   }
   if (L->cursor == NULL || L->index < 0){
    printf("List error: undefined cursor ");
   }

    return L->cursor->data;
}



bool noIndex(List L){
if(index(L) == 0){return true;}
else{
    return false;
}
}
// equals
// returns 1 if this List and L are the
// same integer sequence.
//the cursor is ignored in both lists.
int equals(List A, List B) {

    if( A==NULL || B==NULL )
   {
      printf("List Error: calling equals() on NULL List reference\n");
      return true;
   }
    Node b = B->front;
    Node a = A->front;
    for (;a != NULL && b != NULL; a = a->next, b = b->next) {
        if (a->data != b->data){
            break;
        }

    }
    if (a != NULL || b != NULL){
        return 0;
    }
   return 1;
}

// Manipulation procedures ------------------------------------------------

// clear()
// resets this List to its original empty state
void clear(List L) {
    if (L == NULL){
        printf("List Error: List is NULL \n");
        exit(1);
    }

    if (L->front == NULL && L->back == NULL)
        return;
    while (notEmpty(L)) {

  deleteFront(L);
    }
}

void moveFront(List L){
if (L == NULL)
    {
        printf("List Error: calling moveFront() on a NULL List reference\n");
        exit(1);
    }
//    if (!notEmpty(L))
//    {
//      printf("List Error: calling moveFront() on empty List \n");
//    }
    else{
        L->index = 0;
        L->cursor = L->front;

    }
}
void moveBack(List L){
if (L == NULL)
    {
        printf("List Error: calling moveFront() on a NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
      printf("List Error: calling moveFront() on empty List \n");
    }
    else{

        L->index = L->length - 1;
        L->cursor = L->back;
    }
}
void movePrev(List L){
if (L == NULL)
    {
        printf("List Error: calling moveFront() on a NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
      printf("List Error: calling moveFront() on empty List \n");
    }
    if (L->cursor == L->front){

        L->cursor = NULL;
        L->index = -1;
    }
    else
    {
        L->cursor = L->cursor->prev;
        L->index--;
    }
}
void moveNext(List L){
if (L == NULL)
    {
        printf("List Error: calling moveFront() on a NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
      printf("List Error: calling moveFront() on empty List \n");
    }
    if (L->cursor == L->back){

        L->cursor = NULL;
        L->index = -1;
    }
    else
    {
        L->cursor = L->cursor->next;
        L->index++;
    }
}
void prepend(List L, int data){
Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   if( !notEmpty(L) ) {
      L->front = L->back = N;
   }else{
    if(L->cursor != NULL){
        L->index++;
    }
      L->front->prev = N;
      N->next = L->front;
      L->front = N;
      N->prev = NULL;
   }
   L->length++;
}
void append(List L, int data){

   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   if( !notEmpty(L) ) {
      L->front = L->back = N;
   }else{
      L->back->next = N;
      N->prev = L->back;
      L->back = N;
      N->next = NULL;
   }
   L->length++;

}

void insertBefore(List L, int data){

    if (L == NULL)
    {
        printf("List Error: calling on a NULL List reference\n");
        exit(1);
    }
    if (L->cursor == NULL)
    {
        printf("List Error: cannot insertBefore()  if cursor is undefined\n");
    }
    if (cFront(L))
    {
        prepend(L, data);
    }
    else
    {
        Node N = newNode(data);
        Node pN = L->cursor->prev;

            N->prev = pN;
            N->next = L->cursor;
            pN->next = N;
            L->cursor->prev = N;

            L->length++;
            L->index++;
    }
}

bool cBack(List L){
if(L->cursor == L->back){return true;}
else{
    return false;
}
}

bool curNull(List L){
if(L->cursor == NULL){return true;}
else{
    return false;
}
}

void insertAfter(List L, int data){
    if (L == NULL)
    {
        printf("List Error: a NULL List reference\n");
        exit(1);
    }
    if (!notEmpty(L))
    {
        printf("List Error: insertAfter() called on empty List\n");
    }
    if (cBack(L))
    {
        append(L, data);
    }
    else
    {
        Node N = newNode(data);
        Node nN = L->cursor->next;

            N->next = nN;
            N->prev = L->cursor;
            nN->prev = N;
            L->cursor->next = N;
            L->length++;
    }
}

void deleteFront(List L){
   if( !notEmpty(L) ){
      printf("List error: List already empty \n");

   }
   if(L == NULL){
    printf("List error: Null List Reference");
    exit(1);
   }
    if (length(L) > 1)      {
          Node N = L->front;
          L->front = L->front->next;
          L->front->prev = NULL;
          if (!curNull(L))
          {
              L->index--;
          }
          freeNode(&N);
   }
   else if(length(L) == 1){
        Node N = L->front;
        freeNode(&N);
        L->front = NULL;
        L->back = NULL;
        L->index = -1;
        L->cursor = NULL;

   }
   L->length--;

}

void deleteBack(List L){
    if (L == NULL)
    {
        printf("List Error: cannot deleteBack() on NULL List reference\n");
    }
    if (!notEmpty(L))
    {
        printf("List Error: cannot deleteBack() empty List\n");
    }

        if (L->length > 1)
        {
            if (index(L) == length(L) -1)
            {
                L->index = -1;
            }
            Node N = L->back;
            L->back = L->back->prev;
            L->back->next = NULL;
            freeNode(&N);
        }

        else if (L->length == 1)
        {
            Node N = L->back;
            freeNode(&N);
            L->front = NULL;
            L->back = NULL;
            L->index = -1;
            L->cursor = NULL;
        }
        L->length--;
}

void delete(List L){
    if( !notEmpty(L) ){
        printf("List error: List already empty \n");
        exit(1);
   }
   if(L == NULL){
        printf("List error: Null List Reference");
        exit(1);
   }
    if(L->front == L->cursor){
        deleteFront(L);
    }
    else if(L->back == L->cursor){
        deleteBack(L);
    }
    else{
        if(L->cursor->next != NULL){
            L->cursor->next->prev = L->cursor->prev;
        }
        if(L->cursor->prev != NULL){
            L->cursor->prev->next = L->cursor->next;
        }
        L->cursor = NULL;
        L->length--;
        L->index = -1;
    }


}


void printList(FILE* out, List L){
    Node N = NULL;
    if(L == NULL)
    {
        printf("List error: printList() called on NULL List reference\n");
        exit(1);
    }
    for (N = L->front; N != NULL; N = N->next)
    {
        fprintf(out, "%d", N->data);
        fprintf(out, " ");
    }
}


List copyList(List L){
    List C = newList();
    if (length(L) > 0)
    {
        Node N = L->front;
        while (N != NULL)
        {
            append(C, N->data);
            N = N->next;
        }
    }
    C->cursor = NULL;
    C->index = -1;
    return C;
}
